File: eco-api/src/main/java/com/willfp/eco/internal/gui/MenuHandler.java
package com.willfp.eco.internal.gui;

import com.willfp.eco.core.gui.menu.Menu;
import lombok.experimental.UtilityClass;
import org.bukkit.inventory.Inventory;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.HashMap;
import java.util.Map;

@UtilityClass
public class MenuHandler {
    private static final Map<Inventory, Menu> MENUS = new HashMap<>();

    public void registerMenu(@NotNull final Inventory inventory,
                             @NotNull final Menu menu) {
        MENUS.put(inventory, menu);
    }

    public void unregisterMenu(@NotNull final Inventory inventory) {
        MENUS.remove(inventory);
    }

    @Nullable
    public Menu getMenu(@NotNull final Inventory inventory) {
        return MENUS.get(inventory);
    }
}

File: eco-core/core-backend/src/main/kotlin/com/willfp/eco/internal/gui/menu/MenuHandler.kt
package com.willfp.eco.internal.gui.menu

import com.willfp.eco.core.gui.menu.Menu
import java.util.Collections
import java.util.WeakHashMap
import org.bukkit.entity.Player
import org.bukkit.inventory.Inventory

private val inventories: MutableMap<Inventory, RenderedInventory> =
    Collections.synchronizedMap(WeakHashMap())

object MenuHandler {
    fun registerInventory(
        inventory: Inventory,
        menu: EcoMenu,
        player: Player
    ): RenderedInventory {
        val rendered = RenderedInventory(menu, inventory, player)
        inventories[inventory] = rendered
        return rendered
    }

    fun unregisterInventory(inventory: Inventory) {
        inventories.remove(inventory)
    }
}

fun Inventory.asRenderedInventory(): RenderedInventory? =
    inventories[this]

fun Inventory.getMenu(): Menu? =
    this.asRenderedInventory()?.menu
